### PR TITLE
pause-resume resolved confilicts 

### PR DESCRIPTION
**Pull request by mesat**
pause-resume resolved conflicts with 9d88270 and fcc2338 for Commits on Dec 23, 2019

### CODE REVIEW

**Comment by mesat** [comment]
Because of the changed branch, I created a new pull request.

**Comment by mesat** [comment]
If it's necessary to continue from #56, I can create from the master of the fork

**Comment by ctongfei** [comment]
So we could close #56 ? I'll review this soon.

File: src/main/java/me/tongfei/progressbar/ProgressState.java [hunk 1]
```diff
@@ -67,6 +74,8 @@ synchronized long getMax() {
     // The progress, normalized to range [0, 1].
     synchronized double getNormalizedProgress() {
         if (max <= 0) return 0.0;
+        else if (max - start < current)
```
Comment (ctongfei) [inline src/main/java/me/tongfei/progressbar/ProgressState.java]: I don't follow the logic here -- what condition does `max - start < current` check?

File: src/main/java/me/tongfei/progressbar/ProgressState.java [hunk 2]
```diff
@@ -15,11 +17,16 @@
     long max = 0;
     Instant startTime = null;
     String extraMessage = "";
+    long start = 0;
+    Duration elapsed = Duration.ofSeconds(0L);
 
-    ProgressState(String task, long initialMax) {
+    ProgressState(String task, long initialMax, long startFrom, long elapsedSecond) {
```
Comment (ctongfei) [inline src/main/java/me/tongfei/progressbar/ProgressState.java]: the argument `long elapsedSecond` would be better changed to `Duration elapsed`.

File: src/main/java/me/tongfei/progressbar/ProgressBar.java [hunk 3]
```diff
@@ -83,11 +84,14 @@ public ProgressBar(
             String unitName,
             long unitSize,
             boolean showSpeed,
-            DecimalFormat speedFormat
+            DecimalFormat speedFormat,
+            ChronoUnit speedUnit,
+            long startFrom,
+            long elapsedSecond
```
Comment (ctongfei) [inline src/main/java/me/tongfei/progressbar/ProgressBar.java]: Argument `long elapsedSecond` should be changed to `Duration elapsed` to better reflect the intent.

File: src/main/java/me/tongfei/progressbar/ProgressBar.java [hunk 4]
```diff
@@ -99,20 +103,24 @@ public ProgressBar(
...
      */
     public ProgressBar(
             String task,
             long initialMax,
             int updateIntervalMillis,
             ProgressBarRenderer renderer,
-            ProgressBarConsumer consumer
+            ProgressBarConsumer consumer,
+            long startFrom,
+            long elapsedSecond
```
Comment (ctongfei) [inline src/main/java/me/tongfei/progressbar/ProgressBar.java]: I would prefer moving `long startFrom, Duration elapsed` before the renderer and the consumer. The arguments are grouped into 3 groups semantically: the progress state, the renderer, and the consumer.

File: src/main/java/me/tongfei/progressbar/DefaultProgressBarRenderer.java [hunk 5]
```diff
@@ -64,10 +69,33 @@ private String ratio(ProgressState progress) {
...
 
     private String speed(ProgressState progress, Duration elapsed) {
-        if (elapsed.getSeconds() == 0) return "?" + unitName + "/s";
-        double speed = (double) progress.current / elapsed.getSeconds();
+        String suffix = null;
+        double elapsedWithUnit = 0;
+        if (null != speedUnit)
+            switch (speedUnit) {
+                case MINUTES:
+                    suffix = "/m";
```
Comment (ctongfei) [inline src/main/java/me/tongfei/progressbar/DefaultProgressBarRenderer.java]: `/min` instead of `/m`.

**Pull request merged**